theano/tensor/nnet/nnet.py

```diff
@@ -130,284 +131,574 @@ def c_code_template(dtype):
             inside_row_loop_contig = """
-            for (j = 0; j < Nx[1]; ++j)
+            // Substract the max
+            for (j = 0; j < num_sm_classes; ++j)
             {
-                dtype_%%(sm)s row_ij = x_i[j * Sx] +  b_i[j * Sb];
-                //std::cout << "2 " << j << " " << row_ij << " " << row_max << "\\n";
-                dtype_%%(sm)s sm_ij = row_ij - row_max;
-                //std::cout << "3 " << j << " " << sm_ij << "\\n";
-                sm_i[j * Ssm] = sm_ij;
+                sm_i[j * stride_sm] = (x_i[j * stride_x] + b_i[j * stride_b]) - row_max;
             }
+
+            // Take the exponential
             %(vec_exp)s;
-            for (j = 0; j < Nx[1]; ++j)
+
+            // Compute the sum of exponentials
+            for (j = 0; j < num_sm_classes; ++j)
             {
-                sum += sm_i[j * Ssm];
-            }
+                sum += sm_i[j * stride_sm];
+                }
 
+            // Then normalize each element
             //cblas_dscal(x.N, 1.0 / sum, &mat_at(s,i,0), s.n);
             double sum_inv = 1.0 / sum;
-            for (j = 0; j < Nx[1]; ++j)
+            for (j = 0; j < num_sm_classes; ++j)
             {
-                sm_i[j * Ssm] *= sum_inv;
-            }
+                sm_i[j * stride_sm] *= sum_inv;
+                }
+            """ % locals()
 
-        """ % locals()
             inside_row_loop = """
-            if(Ssm == 1){
+            if(stride_sm == 1){
                 %(inside_row_loop_contig)s
             }else{
                 %(inside_row_loop)s
             }
             """ % locals()
         except theano.gof.utils.MethodNotDefined:
             pass
+
         end_row_loop = """
-        }
+            PyArray_ITER_NEXT(it_x);
+            PyArray_ITER_NEXT(it_sm);
+            }
         """
-
         return (init_decl, begin_row_loop, inside_row_loop, end_row_loop)
 
     def c_code(self, node, name, inp, out, sub):
-        x, b = inp
+        x, b, axis = inp
         sm, = out
         code_template = ''.join(self.c_code_template(
             node.inputs[0].type.dtype_specs()[1]))
         return code_template % dict(locals(), **sub)
 
     @staticmethod
     def c_code_cache_version():
-        return (8,)
+        return (9,)
 
 softmax_with_bias = SoftmaxWithBias()
 
 
 class SoftmaxGrad(gof.Op):
     """
     Gradient wrt x of the Softmax Op.
-
     """
 
-    nin = 2
+    nin = 3
     nout = 1
     __props__ = ()
 
-    def make_node(self, dy, sm):
+    def make_node(self, dy, sm, axis=-1):
         dy = tensor.as_tensor_variable(dy)
         sm = tensor.as_tensor_variable(sm)
-        if dy.type.ndim not in (1, 2) \
-                or dy.type.dtype not in tensor.float_dtypes:
-            raise ValueError('dy must be 1-d or 2-d tensor of floats. Got ',
-                             dy.type)
+        axis = tensor.as_tensor_variable(axis)
```

Comment:
If we want to keep the axis as a node in the graph, we still need to define it as a tensor variable.
Yeah, I will do some cleaning in the code.